kernel/cemitcodes.ml

```diff
@@ -369,10 +369,17 @@ let to_memory (init_code, fun_code, fv) =
   init();
   emit init_code;
   emit fun_code;
-  let code = String.create !out_position in
-  String.unsafe_blit !out_buffer 0 code 0 !out_position;
   (** Later uses of this string are all purely functional *)
-  let code = CString.hcons code in
+  let code = Bytes.init !out_position (fun i ->
+      Bytes.unsafe_get !out_buffer i
+    ) in
+  (* XXX: We lose the hconsing optimization wrt the non-bytes version.
+
+     This really depends on whether we want emitcodes to be `string` or `bytes`.
+
+     The current choice is `bytes`, but `string` can be done using a
+     more careful patching. *)
```

Comment:
Note that `unsafe_to_string`, while unsafe, has a clear specification in terms of ownership: see [the Bytes documentation: Unsafe conversions](http://caml.inria.fr/pub/docs/manual-ocaml/libref/Bytes.html#4_Unsafeconversionsforadvancedusers). In particular, I think it is important to *check* that unsafe can be safely used here (even if you end up not using it) because I think that it is on the same ownership invariant that the correctness of the current code *relies* to not make a copy there. So if you could prove that using `unsafe_to_string` is wrong, you would probably have found a bug.